service/authorization/authorization.go

```diff
@@ -359,40 +360,107 @@ func (as *AuthorizationService) GetDecisions(ctx context.Context, req *authoriza
+			values[i].SubjectMappings = subjectMappings
 		}
-		var attributeFqns []string
-		for _, attr := range listAttributeResp.GetAttributes() {
-			for _, val := range attr.GetValues() {
-				attributeFqns = append(attributeFqns, val.GetFqn())
-			}
+	}
+	return values
+}
+
+// updateValsByFqnLookup updates the lookup map with attribute values by FQN.
+func updateValsByFqnLookup(attribute *policy.Attribute, scopeMap map[string]bool, fqnAttrVals map[string]*attr.GetAttributeValuesByFqnsResponse_AttributeAndValue) map[string]*attr.GetAttributeValuesByFqnsResponse_AttributeAndValue {
+	rule := attribute.GetRule()
+	for _, v := range attribute.GetValues() {
+		// if scope exists and current attribute value FQN is not in scope
+		if !(scopeMap == nil || scopeMap[v.GetFqn()]) {
+			// skip
+			continue
+		}
+		// trim attribute values (by default only keep single value relevant to FQN)
+		// This is key to minimizing the rego query size.
+		values := []*policy.Value{v}
+		if rule == policy.AttributeRuleTypeEnum_ATTRIBUTE_RULE_TYPE_ENUM_HIERARCHY {
+			// restore ALL attribute values if attribute rule is hierarchical
+			// This is key to honoring comprehensive hierarchy.
+			values = attribute.GetValues()
 		}
-		request.Fqns = attributeFqns
-	} else {
-		// get subject mappings
-		request.Fqns = req.GetScope().GetAttributeValueFqns()
+		// only clone relevant fields for attribute
+		a := &policy.Attribute{Rule: rule, Values: values}
+		fqnAttrVals[v.GetFqn()] = &attr.GetAttributeValuesByFqnsResponse_AttributeAndValue{Attribute: a, Value: v}
+	}
+	return fqnAttrVals
+}
+
+// makeValsByFqnsLookup creates a lookup map of attribute values by FQN.
+func makeValsByFqnsLookup(attrs []*policy.Attribute, subMapsByVal map[string][]*policy.SubjectMapping, scopeMap map[string]bool) map[string]*attr.GetAttributeValuesByFqnsResponse_AttributeAndValue {
+	fqnAttrVals := make(map[string]*attr.GetAttributeValuesByFqnsResponse_AttributeAndValue)
+	for i := range attrs {
+		// add subject mappings to attribute values
+		attrs[i].Values = updateValsWithSubMaps(attrs[i].GetValues(), subMapsByVal)
+		// update the lookup map with attribute values by FQN
+		fqnAttrVals = updateValsByFqnLookup(attrs[i], scopeMap, fqnAttrVals)
+	}
+	return fqnAttrVals
+}
+
+// makeScopeMap creates a map of attribute value FQNs.
+func makeScopeMap(scope *authorization.ResourceAttribute) map[string]bool {
+	// if scope not defined, return nil pointer
+	if scope == nil {
+		return nil
+	}
+	scopeMap := make(map[string]bool)
+	// add attribute value FQNs from scope to the map
+	for _, fqn := range scope.GetAttributeValueFqns() {
+		scopeMap[fqn] = true
+	}
+	return scopeMap
+}
+
+func (as *AuthorizationService) GetEntitlements(ctx context.Context, req *authorization.GetEntitlementsRequest) (*authorization.GetEntitlementsResponse, error) {
+	as.logger.DebugContext(ctx, "getting entitlements")
+	attrsRes, err := as.sdk.Attributes.ListAttributes(ctx, &attr.ListAttributesRequest{})
```

Comment:
@suchak1 Do we even need to make this call?

Subject mappings should return the value and its fqn but it seems like the sql query is not returning that field. Maybe this requires a follow up and larger discussion around what is actually returned in these calls.

Reply:
@strantalis I'd prefer to create an issue to add support for ListAttributes to allow returning values as well. That said, returning values for every attribute seems expensive. If we wanted this optimization, we should figure out how to reduce the number of attributes needed in the lookup.

Reply:
I'm a little concerned about `ListAttributes` becoming `GetAllPolicy` which seems like what would be needed for all of this in-memory restructuring. That said, we can tackle the selector enhancements to optionally support it if desired.

Reply:
The sql query was just missing the join statement to get the fqn for the attribute value.

```
{
  "subject_mappings": [
    {
      "id": "7e1237d0-5445-4a8e-95f6-5c1e3bb4a125",
      "attribute_value": {
        "id": "01ea3af5-97e9-40d0-b148-2b061cfe53d3",
        "attribute": null,
        "value": "mar",
        "members": [],
        "grants": [],
        "fqn": "https://demo.com/attr/relto/value/mar",
        "active": true,
        "subject_mappings": [],
        "metadata": null
      },
      "subject_condition_set": {
        "id": "bb7145fa-5cc2-46eb-9d4d-b0d83bee284e",
        "subject_sets": [
          {
            "condition_groups": [
              {
                "conditions": [
                  {
                    "subject_external_selector_value": ".attributes.nationality[]",
                    "operator": "SUBJECT_MAPPING_OPERATOR_ENUM_IN",
                    "subject_external_values": [
                      "MAR"
                    ]
                  },
                  {
                    "subject_external_selector_value": ".clientId",
                    "operator": "SUBJECT_MAPPING_OPERATOR_ENUM_IN",
                    "subject_external_values": [
                      "dsp-outlook",
                      "dsp-sharepoint",
                      "shared",
                      "dsp-shared",
                      "opentdf",
                      "dsp-tdf-reader",
                      "dsp-tdf-reader-web",
                      "dsp-windows-filewatcher"
                    ]
                  }
                ],
                "boolean_operator": "CONDITION_BOOLEAN_TYPE_ENUM_OR"
              }
            ]
          }
        ],
        "metadata": {
          "created_at": "2024-08-08T17:06:31.155605Z",
          "updated_at": "2024-08-08T17:06:31.155605Z",
          "labels": {}
        }
      },
      "actions": [
        {
          "standard": "STANDARD_ACTION_DECRYPT"
        }
      ],
      "metadata": {
        "created_at": "2024-08-08T17:06:31.163358Z",
        "updated_at": "2024-08-08T17:06:31.163358Z",
        "labels": {}
      }
    },
......
```